adapter/src/ethereum.rs

```diff
@@ -1,101 +1,505 @@
+                .map_err(|_| map_error("Invalid keystore json provided"))?,
+            None,
+            &Some(self.keystore_pwd.clone()),
+        )
+        .map_err(|_| map_error("Failed to create account"))?;
+
+        self.wallet.replace(Some(account));
+
+        // wallet has been unlocked
         Ok(true)
     }
 
-    fn whoami(&self) -> String {
-        self.address.unwrap().to_string()
+    fn whoami(&self) -> AdapterResult<String> {
+        match self.wallet.borrow().clone() {
+            Some(wallet) => {
+                let public = wallet
+                    .public(&self.keystore_pwd)
+                    .map_err(|_| map_error("Failed to get public key"))?;
+                let address = format!("{:?}", public_to_address(&public));
+                Ok(eth_checksum::checksum(&address))
+            }
+            None => Err(AdapterError::Configuration(
+                "Unlock wallet before use".to_string(),
+            )),
+        }
     }
 
     fn sign(&self, state_root: &str) -> AdapterResult<String> {
-        let signature = format!(
-            "Dummy adapter signature for {} by {}",
-            state_root,
-            self.whoami()
-        );
-        Ok(signature)
+        let message = Message::from_slice(&hash_message(state_root));
+        match self.wallet.borrow().clone() {
+            Some(wallet) => {
+                let wallet_sign = wallet
+                    .sign(&self.keystore_pwd, &message)
+                    .map_err(|_| map_error("failed to sign messages"))?;
+                let signature: Signature = wallet_sign.into_electrum().into();
+                Ok(format!("0x{}", signature))
+            }
+            None => Err(AdapterError::Configuration(
+                "Unlock the wallet before signing".to_string(),
+            )),
+        }
     }
 
-    fn verify(&self, signer: &str, _state_root: &str, signature: &str) -> AdapterResult<bool> {
-        // select the `identity` and compare it to the signer
-        // for empty string this will return array with 1 element - an empty string `[""]`
-        let is_same = match signature.rsplit(' ').take(1).next() {
-            Some(from) => from == signer,
-            None => false,
+    fn verify(&self, signer: &str, state_root: &str, sig: &str) -> AdapterResult<bool> {
+        let (decoded_adress, decoded_signature) = match (hex::decode(signer), hex::decode(sig)) {
+            (Ok(address), Ok(sig)) => (address, sig),
+            (_, _) => {
+                return Err(AdapterError::Signature(
+                    "invalid signature or address".to_string(),
+                ))
+            }
         };
 
-        Ok(is_same)
+        let address = Address::from_slice(&decoded_adress);
+        let signature = Signature::from_electrum(&decoded_signature);
+        let message = Message::from_slice(&hash_message(state_root));
+
+        match verify_address(&address, &signature, &message) {
```

Comment:
Note for next time: you could use `map_err`, as you are not changing the Ok result, but only the error one.

Reply:
I can't use map_err here because the `Err` arm isn't returning an error but `Ok(false)`

Reply:
yes, my bad.